### DISCUSSION TITLE
CIP 20 - Extensible Hash Precompile - Discussions

### DISCUSSION THREAD

**Discussion by prestwich** [celo-org/celo-proposals#334] [CIP Discussions]
Migrating my comments from the forum:

The goal is to provide access to some common cryptographic hash functions. We do that by adding them all to a single precompile. Right now the precompile supports common SHA3 variants (SHA3-256, SHA3-512, and Keccak-512), as well as blake2s

[Link to CIP](https://github.com/celo-org/celo-proposals/blob/master/CIPs/0020.md)

**Deployment Checklist** 
- [x] Specification
- [X] Implementation
    - [X] celo-blockchain [#1249](https://github.com/celo-org/celo-blockchain/pull/1249)
- [ ] Testing
    - [X] Unit tests
    - [X] Fuzzing w/ smash
    - [X] Benching & Gas assignment 
    - [ ] gas calculation tests
- [x] Pre-deploy
    - [x] Precompile Address assignment in Donut block
- [ ] Verification
    - [ ] Run the test vectors via `eth_call` RPC

**Comment by prestwich** [comment]
I ran benchmarks on our blake2s implementation against golang stdlib sha256:

```
BenchmarkHash64-8        6719373               178 ns/op         360.12 MB/s
BenchmarkHash128-8       3734914               322 ns/op         397.93 MB/s
BenchmarkHash1K-8         542121              2209 ns/op         463.63 MB/s
BenchmarkSha64-8         3136120               382 ns/op         167.56 MB/s
BenchmarkSha128-8        2240745               546 ns/op         234.63 MB/s
BenchmarkSha1K-8          418257              2858 ns/op         358.33 MB/s
```

The blake2s implementation outperforms sha2 in all cases. I'd suggest setting the gas costs equal to sha2 for now

**Comment by weijiekoh** [comment]
Is it possible for this precompile to, in the future, enable support for snark-friendly hash functions like Poseidon?
